docker-entrypoint.sh

```diff
@@ -14,76 +14,222 @@
-    if [[ "$?" -ne 0 ]]; then
-        echo "checking the readiness of oap server"
-        sleep 10
+    if [[ $? -ne 0 ]]; then
+        echo "zookeeper server failed to start in 120 seconds: "
+        cat ${ZK_HOME}/logs/*
+        exit 1
     fi
-done
 
-nc -zv ${OAP_HOST} ${OAP_PORT}
+    cd ${original_pwd}
+}
 
-if [[ "$?" -ne 0 ]]; then
-    echo "oap server failed to start in 120 seconds: "
-    cat ${OAP_LOG_DIR}/*
-    exit 1
-fi
+start_es() {
+    original_pwd=$(pwd)
+
+    cd ${ES_HOME}
 
-echo "oap server is ready for connections"
+    export ES_TMPDIR=`mktemp -d -t elasticsearch.XXXXXXXX`
+    export ES_JAVA_OPTS="-Xms1g -Xmx1g"
 
-for i in `seq 1 12`; do
-    nc -z ${WEBAPP_HOST} ${WEBAPP_PORT}
-    if [[ "$?" -ne 0 ]]; then
-        echo "checking the readiness of web app"
-        sleep 10
+    addgroup -S es \
+        && adduser -S es -G es -s /bin/bash -D \
+        && tar -zxf elasticsearch-oss-6.3.2.tar.gz -C /tmp/ \
+        && mv /tmp/elasticsearch-6.3.2/* . \
+        && chown -R es:es ${ES_HOME} \
+        && chown -R es:es /tmp \
+        && su es ${ES_HOME}/bin/elasticsearch > logs/stdout.log 2>&1 &
+
+    check_tcp localhost 9200 12 10 "waiting for the elasticsearch to be ready"
+
+    if [[ $? -ne 0 ]]; then
+        echo "elasticsearch server failed to start in 120 seconds: "
+        cat ${ES_HOME}/logs/elasticsearch.log ${ES_HOME}/logs/stdout.log
+        exit 1
     fi
-done
 
-nc -zv ${WEBAPP_HOST} ${WEBAPP_PORT}
+    cd ${original_pwd}
+}
+
+start_oap() {
+    local mode=$1
+
+    original_pwd=$(pwd)
+
+    cd ${SW_HOME}/
+
+    if test "${mode}" = "init"; then
+        bash bin/oapService.sh > /dev/null 2>&1 &
+    else
+        bash bin/oapServiceNoInit.sh > /dev/null 2>&1 &
+    fi
+    check_times=30
+    check_interval=10
+
+    check_tcp ${SW_CORE_GRPC_HOST} \
+              ${SW_CORE_GRPC_PORT} \
+              ${check_times} \
+              ${check_interval} \
+              "waiting for the oap server to be ready"
+
+    if [[ $? -ne 0 ]]; then
+        echo "oap server failed to start in ${check_times} * ${check_interval} seconds: "
+        cat ${OAP_LOG_DIR}/*
+        exit 1
+    fi
+
+    echo "oap server is ready for connections"
+
+    cd ${original_pwd}
+}
+
+start_webapp() {
+    original_pwd=$(pwd)
+
+    cd ${SW_HOME}/
+
+    bash bin/webappService.sh > /dev/null 2>&1 &
+    check_tcp ${WEBAPP_HOST} \
+          ${WEBAPP_PORT} \
+          ${check_times} \
+          ${check_interval} \
+          "waiting for the web app to be ready"
+
+    if [[ $? -ne 0 ]]; then
+        echo "web app failed to start in ${check_times} * ${check_interval} seconds"
+        cat ${WEBAPP_LOG_DIR}/*
+        exit 1
+    fi
+
+    echo "web app is ready for connections"
+
+    cd ${original_pwd}
+}
+
+export MODE=${MODE:-standalone}
 
-if [[ "$?" -ne 0 ]]; then
-    echo "web app failed to start in 120 seconds"
-    cat ${WEBAPP_LOG_DIR}/*
-    exit 1
+echo "starting e2e container in mode: ${MODE}"
```

Comment:
Set up a preparation stage here, ask which e2e test case provide a `test-preparation.sh` to do everything they need, then, build to test :)